capsules/src/process_console.rs

```diff
@@ -860,8 +978,131 @@ impl<'a, A: Alarm<'a>, C: ProcessManagementCapability> uart::ReceiveClient
                                 // Backspace, echo and remove last byte
                                 // Note echo is '\b \b' to erase
                                 let _ = self.write_bytes(&['\x08' as u8, ' ' as u8, '\x08' as u8]);
-                                command[index - 1] = '\0' as u8;
+                                command[index - 1] = EOL;
                                 self.command_index.set(index - 1);
+
+                                // Remove last byte from the command in order
+                                // not to permit accumulation of the text
+                                if COMMAND_HISTORY_LEN > 1 {
+                                    self.command_history.map(|cmd_arr| {
+                                        (&mut cmd_arr[0]).delete_last_byte();
+                                    });
+                                }
+                            }
+                        } else if (COMMAND_HISTORY_LEN > 1)
+                            && (read_buf[0] == ESC || self.control_seq_in_progress.get())
+                        {
+                            // Catch the Up and Down arrow keys
+                            if read_buf[0] == ESC {
+                                // Signal that a control sequence has started and capture it
+                                self.control_seq_in_progress.set(true);
```

Comment:
I meant to collaborate with Alex on this since we have a downstream version of process console that handles command history.

I want to share the small state machines that we use down stream that you can either choose to use or not (your call). I just wanted it to be an option. Consider this last minute collaboration.


```rust

/// Start of terminal escape sequence
const CHAR_ESC_START: u8 = b'\x1B'; 

/// The FSM for processing terminal escape sequences
#[derive(Copy, Clone)]
enum EscState {
    /// Most recent character is not part of an escapee sequence, so process it elsewhere
    Bypass,
    /// The escape sequence completed with a recognized escape key
    Complete(EscKey),
    Started,
    Bracket,
    Bracket1,
    Bracket3,
    Bracket4,
    Unrecognized,
    UnrecognizedDone,
}

impl EscState {
    /// Process the data for this state and returns the next state in the FSM
    fn next_state(self, data: u8) -> Self {
        use self::{EscKey::*, EscState::*};
        match (self, data) {
            (Bypass, CHAR_ESC_START)
            | (UnrecognizedDone, CHAR_ESC_START)
            | (Complete(_), CHAR_ESC_START) => Started,
            (Bypass, _) | (UnrecognizedDone, _) | (Complete(_), _) => Bypass,
            (Started, b'[') => Bracket,
            (Bracket, b'A') => Complete(Up),
            (Bracket, b'B') => Complete(Down),
            (Bracket, b'C') => Complete(Right),
            (Bracket, b'D') => Complete(Left),
            (Bracket, b'1') => Bracket1,
            (Bracket1, b'~') => Complete(Home),
            (Bracket, b'3') => Bracket3,
            (Bracket3, b'~') => Complete(Delete),
            (Bracket, b'4') => Bracket4,
            (Bracket4, b'~') => Complete(End),
            _ => {
                if EscState::terminator_esc_char(data) {
                    UnrecognizedDone
                } else {
                    Unrecognized
                }
            }
        }
    }

    /// Returns true if the processing next state logically consumed the data and it should not
    /// be processed elsewhere
    pub fn data_consumed(&self) -> bool {
        !matches!(self, EscState::Bypass)
    }

    fn terminator_esc_char(data: u8) -> bool {
        data.is_ascii_alphabetic() || data == b'~'
    }
}

```

And when it is used, we do something like the follow. Note that `cursor` is the position of the cursor and `len` is the total pending command length.

```rust
        let esc_state = self.esc_state.get().next_state(ch);
        self.esc_state.set(esc_state);

        if let EscState::Complete(key) = esc_state {
            match key {
                EscKey::Right if cursor < len => {
                    Direction::Right.move_cursor(1);
                    cursor += 1;
                }
                EscKey::Left if cursor > 0 => {
                    Direction::Left.move_cursor(1);
                    cursor -= 1;
                }
                EscKey::End if cursor < len => {
                    Direction::Right.move_cursor(len - cursor);
                    cursor = len;
                }
                EscKey::Home if cursor > 0 => {
                    Direction::Left.move_cursor(cursor);
                    cursor = 0;
                }
                EscKey::Delete if cursor < len => {
                    // Remove the character under the cursor
                    try_copy_within(&mut cmd_buf, cursor + 1..len, cursor)?;
                    len -= 1;
                    // Print the updated buffer with a clearing space at end
                    for c in cmd_buf.iter().skip(cursor).take(len - cursor) {
                        console_output_no_eol!("{}", *c as char);
                    }
                    console_output_no_eol!(" ");
                    Direction::Left.move_cursor(len - cursor + 1);
                }
                EscKey::Up | EscKey::Down => {
                    let res: Result<(), ErrorCode> = self.mru.map_or(Ok(()), |mru| {
                        // Clear the current line on console
                        Direction::Left.move_cursor(cursor);
                        for _ in 0..len {
                            console_output_no_eol!(" ");
                        }
                        Direction::Left.move_cursor(len);

                        // Copy in the next data from mru circular buffer slices
                        let (b1, b2) = if matches!(key, EscKey::Up) {
                            mru.get_next_older()
                        } else {
                            mru.get_next_newer()
                        };
                        let total = b1.len() + b2.len();
                        cmd_buf
                            .get_mut(..b1.len())
                            .ok_or(ErrorCode::INVAL)?
                            .copy_from_slice(b1);
                        cmd_buf
                            .get_mut(b1.len()..total)
                            .ok_or(ErrorCode::INVAL)?
                            .copy_from_slice(b2);
                        cursor = total;
                        len = total;

                        // Print all new characters
                        for c in cmd_buf.iter().take(len) {
                            console_output_no_eol!("{}", *c as char);
                        }
                        Ok(())
                    });
                    res?
                }
                _ => {}
            };
        }
        // If the escape sequence consumed the character, we are done processing this character
        if esc_state.data_consumed() {
            return Ok(());
        }
```

Also here is the Direction helper
```rust
/// Direction to move a cursor
enum Direction {
    Left,
    Right,
}

impl Direction {
    fn move_cursor(&self, count: usize) {
        if count != 0 {
            console_output_no_eol!(
                "\x1b[{}{}",
                count,
                match self {
                    Direction::Left => "D",
                    Direction::Right => "C",
                }
            );
        }
    }
}
```

Lastly, we have a Most Recently Used (`mru` in code above) circular buffer cache for console history that compactly stores previous commands with an upper bound limit on the totally number of bytes and stored commands, which ever comes first. I can also share this too with its unit tests.

Also don't feel like you have to make these changes before landing this PR. If you want to incorporate this feedback in a follow up PR, I would be happy to help out. I just wanted to start talking about this.

Reply:
I agree, @Matrix22 @CosminGGeorgescu you can prepare a new pull request as soon as this gets merged.